expression/util.go

```diff
@@ -905,20 +905,33 @@ func ContainCorrelatedColumn(exprs []Expression) bool {
 	return false
 }
 
-// ContainMutableConst checks if the expressions contain a lazy constant.
-func ContainMutableConst(ctx sessionctx.Context, exprs []Expression) bool {
-	// Treat all constants immutable if plan cache is not enabled for this query.
+// MaybeOverOptimized4PlanCache used to check whether an optimization can work
+// for the statement when we enable the plan cache.
+// In some situations, some optimizations maybe over-optimize and cache an
+// overOptimized plan. The cached plan may not get the correct result when we
+// reuse the plan for other statements. So we need to do the check here.
+// The check includes the following aspects:
+// 1. Whether the plan cache switch is enable.
+// 2. Whether the expressions contain a lazy constant.
+// TODO: Do more careful check here.
+func MaybeOverOptimized4PlanCache(ctx sessionctx.Context, exprs []Expression) bool {
```

Comment:
How about adding a simple case to explain what is `over-optimized`?
For example, `c>=$a and c<=$b` --> `c==$a` ($b is abandoned after optimization)